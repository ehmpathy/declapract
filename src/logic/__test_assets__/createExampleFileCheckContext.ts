import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
