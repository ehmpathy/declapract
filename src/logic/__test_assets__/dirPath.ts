export const testAssetsDirectoryPath = __dirname;
