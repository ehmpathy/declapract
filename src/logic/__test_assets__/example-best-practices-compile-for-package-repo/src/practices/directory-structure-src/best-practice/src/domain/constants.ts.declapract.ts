export const check = { optional: true }; // some projects don't have constants, and thats fine; this shows what naming should look like though
