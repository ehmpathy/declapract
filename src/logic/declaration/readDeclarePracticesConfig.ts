import { config } from 'shelljs';
import { ActionDeclarePracticesConfigInput } from '../../domain/objects/ActionDeclarePracticesConfigInput';
import { DeclaredPractices } from '../../domain/objects/DeclaredPractices';
import { readYmlFile } from '../../utils/fileio/readYmlFile';
import { getDirOfPath } from '../../utils/filepaths/getDirOfPath';
import { UserInputError } from '../UserInputError';
import { readExampleDeclarations } from './readPracticeDeclarations/readExampleDeclarations';
import { readPracticeDeclarations } from './readPracticeDeclarations/readPracticeDeclarations';
import { readUseCaseDeclarations } from './readPracticeDeclarations/readUseCaseDeclarations';

export const readDeclarePracticesConfig = async ({
  configPath,
}: {
  configPath: string;
}): Promise<DeclaredPractices> => {
  const configDir = getDirOfPath(configPath);
  const getAbsolutePathFromRelativeToConfigPath = (relpath: string) => `${configDir}/${relpath}`;

  // get the yml
  const contents = await (async () => {
    try {
      return await readYmlFile({ filePath: configPath });
    } catch (error) {
      throw new UserInputError(`could not read config. ${error.message}. See '${configPath}'`);
    }
  })();

  // validate it into an input object
  const configInput = new ActionDeclarePracticesConfigInput(contents); // applies runtime validation

  // define the practices based on the input
  const practices = await readPracticeDeclarations({
    declaredPracticesDirectory: getAbsolutePathFromRelativeToConfigPath(configInput.declare.practices),
  });

  // define the examples based on the input
  const examples = configInput.declare.examples
    ? await readExampleDeclarations({
        declarationsRootDirectory: configDir,
        declaredExamplesDirectory: getAbsolutePathFromRelativeToConfigPath(configInput.declare.examples),
      })
    : [];

  // define the use cases based on the input
  const useCases = await readUseCaseDeclarations({
    declaredUseCasesPath: getAbsolutePathFromRelativeToConfigPath(configInput.declare['use-cases']),
    practices,
    examples,
  });

  // return the config
  return new DeclaredPractices({
    rootDir: configDir,
    practices,
    useCases,
    examples,
  });
};
