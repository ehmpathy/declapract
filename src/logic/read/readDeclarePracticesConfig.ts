import { ActionDeclarePracticesConfigInput } from '../../domain/objects/ActionDeclarePracticesConfigInput';
import { DeclaredPractices } from '../../domain/objects/DeclaredPractices';
import { readYmlFile } from '../../utils/fileio/readYmlFile';
import { getDirOfPath } from '../../utils/filepaths/getDirOfPath';
import { UserInputError } from '../UserInputError';
import { readExampleDeclarations } from './readDeclarations/readExampleDeclarations';
import { readPracticeDeclarations } from './readDeclarations/readPracticeDeclarations';
import { readUseCaseDeclarations } from './readDeclarations/readUseCaseDeclarations';

export const readDeclarePracticesConfig = async ({
  configPath,
}: {
  configPath: string;
}): Promise<DeclaredPractices> => {
  const configDir = getDirOfPath(configPath);
  const getAbsolutePathFromRelativeToConfigPath = (relpath: string) => `${configDir}/${relpath}`;

  // get the yml
  const contents = await (async () => {
    try {
      return await readYmlFile({ filePath: configPath });
    } catch (error) {
      throw new UserInputError(`could not read config. ${error.message}. See '${configPath}'`);
    }
  })();

  // validate it into an input object
  const configInput = new ActionDeclarePracticesConfigInput(contents); // applies runtime validation

  // define the practices based on the input
  const practices = await readPracticeDeclarations({
    declaredPracticesDirectory: getAbsolutePathFromRelativeToConfigPath(configInput.declare.practices),
  });

  // define the use cases based on the input
  const useCases = await readUseCaseDeclarations({
    declaredUseCasesPath: getAbsolutePathFromRelativeToConfigPath(configInput.declare['use-cases']),
    practices,
  });

  // define the examples based on the input
  const examples = configInput.declare.examples
    ? await readExampleDeclarations({
        declaredExamplesDirectory: getAbsolutePathFromRelativeToConfigPath(configInput.declare.examples),
      })
    : [];

  // return the config
  return new DeclaredPractices({
    rootDir: configDir,
    practices,
    useCases,
    examples,
  });
};
