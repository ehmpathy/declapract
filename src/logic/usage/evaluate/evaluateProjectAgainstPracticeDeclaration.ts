import chalk from 'chalk';

import {
  FileCheckPurpose,
  FileEvaluationResult,
  hasPassed,
  PracticeDeclaration,
  ProjectVariablesImplementation,
} from '../../../domain';
import { FilePracticeEvaluation } from '../../../domain/objects/FilePracticeEvaluation';
import { evaluteProjectAgainstProjectCheckDeclaration } from './evaluateProjectAgainstProjectCheckDeclaration';

export const evaluteProjectAgainstPracticeDeclaration = async ({
  practice,
  projectRootDirectory,
  projectVariables,
}: {
  practice: PracticeDeclaration;
  projectRootDirectory: string;
  projectVariables: ProjectVariablesImplementation;
}) => {
  // evaluate all of the best practices checks and bad practices checks
  const bestPracticeFileCheckEvaluations = practice.bestPractice
    ? await evaluteProjectAgainstProjectCheckDeclaration({
        practiceRef: `${practice.name}.best-practice`,
        purpose: FileCheckPurpose.BEST_PRACTICE,
        projectRootDirectory,
        declaration: practice.bestPractice,
        projectVariables,
      })
    : [];
  const badPracticeFileCheckEvaluations = (
    await Promise.all(
      practice.badPractices.map((badPractice) =>
        evaluteProjectAgainstProjectCheckDeclaration({
          practiceRef: `${practice.name}.bad-practice.${badPractice.name}`,
          purpose: FileCheckPurpose.BAD_PRACTICE,
          projectRootDirectory,
          declaration: badPractice,
          projectVariables,
        }),
      ),
    )
  ).flat();

  // merge the results per file (i.e., if a file passes best practices (or does not exist there) but fails bad practices, just report it as failed once... but if it exists in either and passes all, then report as passed once)
  const pathsEvaluated = [
    ...new Set(
      [
        ...bestPracticeFileCheckEvaluations,
        ...badPracticeFileCheckEvaluations,
      ].map((evaluation) => evaluation.path),
    ),
  ].sort();
  const evaluations = pathsEvaluated.map((path) => {
    const checks = [
      ...bestPracticeFileCheckEvaluations,
      ...badPracticeFileCheckEvaluations,
    ].filter((check) => check.path === path);
    const result = checks.every(hasPassed)
      ? FileEvaluationResult.PASS
      : FileEvaluationResult.FAIL;
    return new FilePracticeEvaluation({
      path,
      result,
      checks,
      practice,
    });
  });
  console.log(chalk.gray(`  âœ“ evaluated practice:${practice.name}`)); // tslint:disable-line: no-console
  return evaluations;
};
