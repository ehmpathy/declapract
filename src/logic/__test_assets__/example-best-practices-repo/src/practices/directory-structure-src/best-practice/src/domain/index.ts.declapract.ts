import { FileCheckType } from '../../../../../../../../../domain';

export const check = FileCheckType.EXISTS;
