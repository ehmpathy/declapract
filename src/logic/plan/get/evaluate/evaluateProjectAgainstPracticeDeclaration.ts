import { FileCheckContext, FileEvaluationResult, hasPassed, PracticeDeclaration } from '../../../../domain';
import { FilePracticeEvaluation } from '../../../../domain/objects/FilePracticeEvaluation';
import { evaluteProjectAgainstProjectCheckDeclaration } from './evaluateProjectAgainstProjectCheckDeclaration';

export const evaluteProjectAgainstPracticeDeclaration = async ({
  practice,
  projectRootDirectory,
}: {
  practice: PracticeDeclaration;
  projectRootDirectory: string;
}) => {
  // evaluate all of the best practices checks and bad practices checks
  const bestPracticeFileCheckEvaluations = practice.bestPractice
    ? await evaluteProjectAgainstProjectCheckDeclaration({
        practiceRef: `${practice.name}.best-practice`,
        context: FileCheckContext.BEST_PRACTICE,
        projectRootDirectory,
        declaration: practice.bestPractice,
      })
    : [];
  const badPracticeFileCheckEvaluations = (
    await Promise.all(
      practice.badPractices.map((badPractice) =>
        evaluteProjectAgainstProjectCheckDeclaration({
          practiceRef: `${practice.name}.bad-practice.${badPractice.name}`,
          context: FileCheckContext.BAD_PRACTICE,
          projectRootDirectory,
          declaration: badPractice,
        }),
      ),
    )
  ).flat();

  // merge the results per file (i.e., if a file passes best practices (or does not exist there) but fails bad practices, just report it as failed once... but if it exists in either and passes all, then report as passed once)
  const pathsEvaluated = [
    ...new Set(
      [...bestPracticeFileCheckEvaluations, ...badPracticeFileCheckEvaluations].map((evaluation) => evaluation.path),
    ),
  ].sort();
  const evaluations = pathsEvaluated.map((path) => {
    const checks = [...bestPracticeFileCheckEvaluations, ...badPracticeFileCheckEvaluations].filter(
      (check) => check.path === path,
    );
    const result = checks.every(hasPassed) ? FileEvaluationResult.PASS : FileEvaluationResult.FAIL;
    return new FilePracticeEvaluation({
      path,
      result,
      checks,
      practice,
    });
  });
  return evaluations;
};
