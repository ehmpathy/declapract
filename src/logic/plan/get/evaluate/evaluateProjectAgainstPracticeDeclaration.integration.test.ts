import { FileEvaluationResult, FileCheckType } from '../../../../domain';
import { readDeclarePracticesConfig } from '../../../config/readDeclarePracticesConfig';
import { testAssetsDirectoryPath } from '../../../__test_assets__/dirPath';
import { evaluteProjectAgainstPracticeDeclaration } from './evaluateProjectAgainstPracticeDeclaration';

describe('evaluteProjectAgainstPracticeDeclaration', () => {
  it('should be able to evaluate a practice with only a best practice', async () => {
    // lookup a practice
    const declarations = await readDeclarePracticesConfig({
      configPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    const practice = declarations.practices.find((practice) => practice.name === 'prettier'); // lets use the "prettier" practice for this one, since its a "best-practice" only one
    if (!practice) fail('should have found a practice');

    // sanity check the practice we'll be using
    expect(practice.badPractices.length).toEqual(0); // check that our expectations for the test are met; should not have any bad practices

    // now evaluate it
    const projectRootDirectory = `${testAssetsDirectoryPath}/example-project-fails-prettier`;
    const evaluations = await evaluteProjectAgainstPracticeDeclaration({ practice, projectRootDirectory });

    // check that the evaluation matches what we expect
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.FAIL).length).toEqual(2); // 2 out of 3 fail
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.PASS).length).toEqual(1); // 1 out of 3 pass
    expect(evaluations).toMatchSnapshot();
  });
  it('should be able to evaluate a practice with both a best practice and bad practices - and consider passing bad practice checks as failing the practice', async () => {
    // lookup a practice
    const declarations = await readDeclarePracticesConfig({
      configPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    const practice = declarations.practices.find((practice) => practice.name === 'dates-and-times');
    if (!practice) fail('should have found a practice');

    // sanity check the practice we'll be using
    expect(practice.badPractices.length).toBeGreaterThan(0); // check that our expectations for the test are met

    // now evaluate it
    const projectRootDirectory = `${testAssetsDirectoryPath}/example-project-fails-dates-and-times`;
    const evaluations = await evaluteProjectAgainstPracticeDeclaration({ practice, projectRootDirectory });

    // check that the evaluation matches what we expect
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.FAIL).length).toEqual(1);
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.PASS).length).toEqual(0);
    expect(evaluations[0].checked.bestPractice[0].result).toEqual(FileEvaluationResult.PASS);
    expect(evaluations[0].checked.badPractices[0].result).toEqual(FileEvaluationResult.PASS);
    expect(evaluations).toMatchSnapshot();
  });
  it('should be able to evaluate a practice with wildcard glob pattern path file checks - fails best practices and passes bad practices', async () => {
    // lookup a practice
    const declarations = await readDeclarePracticesConfig({
      configPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    const practice = declarations.practices.find((practice) => practice.name === 'directory-structure-src');
    if (!practice) fail('should have found a practice');

    // sanity check the practice we'll be using
    expect(practice.badPractices.length).toBeGreaterThan(0); // check that our expectations for the test are met

    // now evaluate it
    const projectRootDirectory = `${testAssetsDirectoryPath}/example-project-fails-directory-structure-src`;
    const evaluations = await evaluteProjectAgainstPracticeDeclaration({ practice, projectRootDirectory });

    // check that the evaluation matches what we expect
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.FAIL).length).toEqual(7);
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.PASS).length).toEqual(2);

    // sanity check a couple of important texamples
    expect(
      evaluations.find((file) => file.path === 'src/logic/**/*.ts'), // note that the file path is still wildcard, since no files were found by that glob
    ).toMatchObject({
      // should have found this file by wildcard _and_ failed it due to the contains check
      result: FileEvaluationResult.FAIL,
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
      }),
    });
    expect(evaluations.find((file) => file.path === 'src/data/clients/coolServiceClient.ts')).toMatchObject({
      // should have found this file by wildcard _and_ failed it due to the contains check not being satisfied correctly
      result: FileEvaluationResult.FAIL,
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({ type: FileCheckType.CONTAINS }),
          }),
        ]),
      }),
    });
    expect(
      evaluations.find((file) => file.path === 'src/services/someFile.ts'), // shold have found this file from the bad-practices wildcard glob path
    ).toMatchObject({
      result: FileEvaluationResult.FAIL, // it failed,
      checked: expect.objectContaining({
        bestPractice: [], // no failed best practices
        badPractices: expect.arrayContaining([
          expect.objectContaining({
            result: FileEvaluationResult.PASS, // the fact that it passed bad practice is why it failed
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
      }),
    });

    // now just save an example of the results
    expect(evaluations).toMatchSnapshot();
  });
  it('should be able to evaluate a practice with wildcard glob pattern path file checks - passes without optionals', async () => {
    // lookup a practice
    const declarations = await readDeclarePracticesConfig({
      configPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    const practice = declarations.practices.find((practice) => practice.name === 'directory-structure-src');
    if (!practice) fail('should have found a practice');

    // sanity check the practice we'll be using
    expect(practice.badPractices.length).toBeGreaterThan(0); // check that our expectations for the test are met

    // now evaluate it
    const projectRootDirectory = `${testAssetsDirectoryPath}/example-project-passes-directory-structure-src-without-optionals`;
    const evaluations = await evaluteProjectAgainstPracticeDeclaration({ practice, projectRootDirectory });
    // console.log(JSON.stringify(evaluation, null, 2));

    // check that the evaluation matches what we expect
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.FAIL).length).toEqual(0);
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.PASS).length).toEqual(9);

    // sanity check a couple of important texamples
    expect(
      evaluations.find((file) => file.path === 'src/contract/handlers/doSomething.ts'), // found this file by wildcard glob path
    ).toMatchObject({
      result: FileEvaluationResult.PASS, // passed, due to existance
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
        badPractices: [],
      }),
    });
    expect(
      evaluations.find((file) => file.path === 'src/data/clients/**/*.ts'), // did not find any files for this glob path -> glob path is kept
    ).toMatchObject({
      result: FileEvaluationResult.PASS,
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({
              required: false, // optional file -> that's why it passed
              type: FileCheckType.CONTAINS,
            }),
          }),
        ]),
      }),
    });
    expect(
      evaluations.find((file) => file.path === 'src/model/**/*'), // should not have found any files in the model dir -> glob path is kept in the check
    ).toMatchObject({
      result: FileEvaluationResult.PASS, // it passed, because the bad practice check failed
      checked: expect.objectContaining({
        bestPractice: [], // no failed best practices
        badPractices: expect.arrayContaining([
          expect.objectContaining({
            result: FileEvaluationResult.FAIL, // the fact that it failed bad practice is why it passed
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
      }),
    });

    // now just save an example of the results
    expect(evaluations).toMatchSnapshot();
  });
  it('should be able to evaluate a practice with wildcard glob pattern path file checks - passes with optionals', async () => {
    // lookup a practice
    const declarations = await readDeclarePracticesConfig({
      configPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    const practice = declarations.practices.find((practice) => practice.name === 'directory-structure-src');
    if (!practice) fail('should have found a practice');

    // sanity check the practice we'll be using
    expect(practice.badPractices.length).toBeGreaterThan(0); // check that our expectations for the test are met

    // now evaluate it
    const projectRootDirectory = `${testAssetsDirectoryPath}/example-project-passes-directory-structure-src-with-optionals`;
    const evaluations = await evaluteProjectAgainstPracticeDeclaration({ practice, projectRootDirectory });
    // console.log(JSON.stringify(evaluation, null, 2));

    // check that the evaluation matches what we expect
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.FAIL).length).toEqual(0);
    expect(evaluations.filter((file) => file.result === FileEvaluationResult.PASS).length).toEqual(9);

    // sanity check a couple of important examples
    expect(
      evaluations.find((file) => file.path === 'src/data/clients/svcAwesomeStuff.ts'), // did not find any files for this glob path -> glob path is kept
    ).toMatchObject({
      result: FileEvaluationResult.PASS,
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({
              required: false, // optional file -> that's why it passed
              type: FileCheckType.CONTAINS,
            }),
          }),
        ]),
      }),
    });
    expect(
      evaluations.find((file) => file.path === 'src/data/dao/superCoolThingDao/index.ts'), // found this file by wildcard glob path
    ).toMatchObject({
      result: FileEvaluationResult.PASS, // passed, due to existance
      checked: expect.objectContaining({
        bestPractice: expect.arrayContaining([
          expect.objectContaining({
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
        badPractices: [],
      }),
    });
    expect(
      evaluations.find((file) => file.path === 'src/model/**/*'), // should not have found any files in the model dir -> glob path is kept in the check
    ).toMatchObject({
      result: FileEvaluationResult.PASS, // it passed, because the bad practice check failed
      checked: expect.objectContaining({
        bestPractice: [], // no failed best practices
        badPractices: expect.arrayContaining([
          expect.objectContaining({
            result: FileEvaluationResult.FAIL, // the fact that it failed bad practice is why it passed
            check: expect.objectContaining({ type: FileCheckType.EXISTS }),
          }),
        ]),
      }),
    });

    // now just save an example of the results
    expect(evaluations).toMatchSnapshot();
  });
});
