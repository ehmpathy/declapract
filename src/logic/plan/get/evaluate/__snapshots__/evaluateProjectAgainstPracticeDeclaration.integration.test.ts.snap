// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluteProjectAgainstPracticeDeclaration should be able to evaluate a practice with both a best practice and bad practices - and consider passing bad practice checks as failing the practice 1`] = `
Array [
  FilePracticeEvaluation {
    "checked": Object {
      "badPractices": Array [
        FileCheckEvaluation {
          "fix": null,
          "path": "package.json",
          "practiceRef": "dates-and-times.bad-practice.moment",
          "reason": null,
          "result": "PASS",
        },
      ],
      "bestPractice": Array [
        FileCheckEvaluation {
          "fix": null,
          "path": "package.json",
          "practiceRef": "dates-and-times.best-practice",
          "reason": null,
          "result": "PASS",
        },
      ],
    },
    "path": "package.json",
    "practice": PracticeDeclaration {
      "badPractices": Array [
        ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": null,
              "pathGlob": "package.json",
              "required": true,
              "type": "CUSTOM",
            },
          ],
          "name": "moment",
          "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
        },
      ],
      "bestPractice": ProjectCheckDeclaration {
        "checks": Array [
          FileCheckDeclaration {
            "check": [Function],
            "fix": null,
            "pathGlob": "package.json",
            "required": true,
            "type": "CUSTOM",
          },
        ],
        "name": "dates-and-times",
        "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
      },
      "name": "dates-and-times",
    },
    "result": "FAIL",
  },
]
`;

exports[`evaluteProjectAgainstPracticeDeclaration should be able to evaluate a practice with only a best practice 1`] = `
Array [
  FilePracticeEvaluation {
    "checked": Object {
      "badPractices": Array [],
      "bestPractice": Array [
        FileCheckEvaluation {
          "fix": [Function],
          "path": ".prettierignore",
          "practiceRef": "prettier.best-practice",
          "reason": "[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

[32m- Expected  - 1[39m
[31m+ Received  + 1[39m

[32m- *.md[39m
[31m+ .js[39m
[2m  â†µ[22m",
          "result": "FAIL",
        },
      ],
    },
    "path": ".prettierignore",
    "practice": PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": ProjectCheckDeclaration {
        "checks": Array [
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": ".prettierignore",
            "required": true,
            "type": "EQUALS",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": null,
            "pathGlob": "package.json",
            "required": true,
            "type": "CUSTOM",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": "prettier.config.js",
            "required": true,
            "type": "EQUALS",
          },
        ],
        "name": "prettier",
        "readme": null,
      },
      "name": "prettier",
    },
    "result": "FAIL",
  },
  FilePracticeEvaluation {
    "checked": Object {
      "badPractices": Array [],
      "bestPractice": Array [
        FileCheckEvaluation {
          "fix": null,
          "path": "package.json",
          "practiceRef": "prettier.best-practice",
          "reason": null,
          "result": "PASS",
        },
      ],
    },
    "path": "package.json",
    "practice": PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": ProjectCheckDeclaration {
        "checks": Array [
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": ".prettierignore",
            "required": true,
            "type": "EQUALS",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": null,
            "pathGlob": "package.json",
            "required": true,
            "type": "CUSTOM",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": "prettier.config.js",
            "required": true,
            "type": "EQUALS",
          },
        ],
        "name": "prettier",
        "readme": null,
      },
      "name": "prettier",
    },
    "result": "PASS",
  },
  FilePracticeEvaluation {
    "checked": Object {
      "badPractices": Array [],
      "bestPractice": Array [
        FileCheckEvaluation {
          "fix": [Function],
          "path": "prettier.config.js",
          "practiceRef": "prettier.best-practice",
          "reason": "[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

[32m- Expected  - 1[39m
[31m+ Received  + 1[39m

[2m  // ref: http://json.schemastore.org/prettierrc[22m

[2m  module.exports = {[22m
[2m    trailingComma: 'all',[22m
[32m-   tabWidth: [7m2[27m,[39m
[31m+   tabWidth: [7m4[27m,[39m
[2m    singleQuote: true,[22m
[2m    printWidth: 150,[22m
[2m  };[22m
[2m  â†µ[22m",
          "result": "FAIL",
        },
      ],
    },
    "path": "prettier.config.js",
    "practice": PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": ProjectCheckDeclaration {
        "checks": Array [
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": ".prettierignore",
            "required": true,
            "type": "EQUALS",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": null,
            "pathGlob": "package.json",
            "required": true,
            "type": "CUSTOM",
          },
          FileCheckDeclaration {
            "check": [Function],
            "fix": [Function],
            "pathGlob": "prettier.config.js",
            "required": true,
            "type": "EQUALS",
          },
        ],
        "name": "prettier",
        "readme": null,
      },
      "name": "prettier",
    },
    "result": "FAIL",
  },
]
`;
